cpp/docs/man/man3/syscalls.3
.TH "syscalls" 3 "SB++" \" -*- nroff -*-
.ad l
.nh
.SH NAME
syscalls \- Syscall Filter Generation This header includes the \fBfilter()\fP command for generating a SECCOMP-BPF filter to restrict syscall within the sandbox\&. Moving to C/C++ allows us to use the libseccomp library directory, which is already installed on any systemd system, allowing us to avoid a dependency where the python bindings are packaged separately\&. Implementation wise, the groups have been removed, as have numerical syscall numbers\&. Groups were a bad idea from the start; any system that combines syscalls will introduce more attack surface, especially for a grouping to be broad enough that a filter can be made by hand--the original goal\&. This also means that offering syscalls on the command line are not supported\&. Secondly, while portability of the application folders isn't a must, numerical syscalls make the files nebulous and system-specific\&.  

.SH SYNOPSIS
.br
.PP
.SS "Functions"

.in +1c
.ti -1c
.RI "std::string \fBfilter\fP (const std::string &application)"
.br
.RI "Generate a syscall filter\&. "
.ti -1c
.RI "void \fBupdate_policy\fP (const std::string &application, const shared::vector &straced)"
.br
.RI "Append an existing policy with newly discovered syscalls\&. "
.in -1c
.SH "Detailed Description"
.PP 
Syscall Filter Generation This header includes the \fBfilter()\fP command for generating a SECCOMP-BPF filter to restrict syscall within the sandbox\&. Moving to C/C++ allows us to use the libseccomp library directory, which is already installed on any systemd system, allowing us to avoid a dependency where the python bindings are packaged separately\&. Implementation wise, the groups have been removed, as have numerical syscall numbers\&. Groups were a bad idea from the start; any system that combines syscalls will introduce more attack surface, especially for a grouping to be broad enough that a filter can be made by hand--the original goal\&. This also means that offering syscalls on the command line are not supported\&. Secondly, while portability of the application folders isn't a must, numerical syscalls make the files nebulous and system-specific\&. 
.SH "Function Documentation"
.PP 
.SS "std::string syscalls::filter (const std::string & application)"

.PP
Generate a syscall filter\&. 
.PP
\fBParameters\fP
.RS 4
\fIapplication\fP The application\&. 
.RE
.PP
\fBReturns\fP
.RS 4
The path to the BPF Filter\&. 
.RE
.PP

.SS "void syscalls::update_policy (const std::string & application, const shared::vector & straced)"

.PP
Append an existing policy with newly discovered syscalls\&. 
.PP
\fBParameters\fP
.RS 4
\fIapplication\fP The name of the application, to locate the syscalls\&.txt 
.br
\fIstrace\fP The output of an straced-run\&. 
.RE
.PP
\fBNote\fP
.RS 4
This function subsumes sb-seccomp in a far better way\&. Rather than parsing the Audit logs (Requiring root permission), we just call --seccomp=strace which will run the program with strace, capture STDERR (The channel strace logs to), and then parse that output to update the file\&. 
.RE
.PP

.SH "Author"
.PP 
Generated automatically by Doxygen for SB++ from the source code\&.


cpp/src/syscalls.cpp
#include "syscalls.hpp"
#include "arguments.hpp"

#include <seccomp.h>
#include <fstream>
#include <cassert>

using namespace shared;
using namespace exec;

namespace syscalls {

  // Create a filter.
  std::string filter(const std::string& application) {

    // Don't make a filter if we aren't supposed to.
    if (arg::at("seccomp") < "permissive") return "";

    // Get our files.
    auto syscall_file = app_data / "syscalls.txt",
    bpf = app_data / "filter.bpf";

    // If we are lacking a syscall file, permissive can just create one, but there's
    // nothing we can do except panic if we're expected to enforce something that doesn't
    // exist.
    if (!std::filesystem::exists(syscall_file)) {
      if (arg::get("seccomp") == "enforcing")
        throw std::runtime_error("Cannot enforce a non-existent syscall policy!");

      // Otherwise we're in permissive mode, so just make the file.
      else {
        log({"Creating syscalls.txt file, check logs to populate!"});
        std::ofstream file(syscall_file);
      }
    }

    // See if the existing file is already up to date, and just use that directly.
    auto content = file::parse<vector>(syscall_file, vectorize);
    if (content.size() > 1) {
      auto hash = std::string(arg::get("seccomp") == "enforcing" ? "E" : "P") + shared::hash(content[1]);
      if (
        hash == container::init<vector>(container::split<vector, char>, content[0], ' ', false)[1] &&
        std::filesystem::exists(bpf) && arg::at("update") < "cache" &&
        arg::get("shell") != "debug" && arg::at("verbose") < "error" &&
        !arg::at("stats")
      ) {
        log({"Using cached SECCOMP filter"});
        return bpf;
      }
    }
    else if (content.size() == 1) {
      if (content[0].starts_with("HASH:")) content = {"0", ""};
      else content = {"0", content[0]};
    }
    else content = {"0", ""};

    // Get the syscalls that should be allowed
    set syscalls = container::init<set>(container::split<set, char>, content[1], ' ', false);
    log({arg::get("seccomp") == "enforcing" ? "Enforced": "Logged", "Syscalls:", std::to_string(syscalls.size())});

    // Add necessary syscalls
    if (arg::get("shell") == "debug")
      extend(syscalls, {"getpgrp", "pselect6"});
    if (arg::at("verbose") >= "error")
      extend(syscalls, {"ptrace", "kill"});

    // Setup the filter.
    auto filter = seccomp_init(arg::get("seccomp") == "enforcing" ? SCMP_ACT_ERRNO(EPERM) : SCMP_ACT_LOG);
    if (filter == nullptr)
      throw std::runtime_error("Failed to initialize SECCOMP!");
    if (seccomp_api_get() < 3)
      throw std::runtime_error("SECCOMP API is too old! Update your kernel!");

    // Set all the settings we want.
    assert(seccomp_attr_set(filter, SCMP_FLTATR_CTL_NNP, true) == 0);
    assert( seccomp_attr_set(filter, SCMP_FLTATR_CTL_TSYNC, true) == 0);
    assert(seccomp_attr_set(filter, SCMP_FLTATR_CTL_OPTIMIZE, 2) == 0);
    if (arg::at("verbose"))
      assert(seccomp_attr_set(filter, SCMP_FLTATR_CTL_LOG, true) == 0);

    // go through each syscall and add it
    for (const auto& syscall : syscalls) {
      auto resolved = seccomp_syscall_resolve_name(syscall.c_str());
      if (resolved < 0) throw std::runtime_error("Unrecognized syscall: "  + syscall);
      seccomp_rule_add(filter, SCMP_ACT_ALLOW, resolved, 0);
    }

    // Write out.
    auto bpf_file = fopen(bpf.c_str(), "w");
    if (bpf_file == nullptr)
      throw std::runtime_error("Failed to open BPF file!");
    seccomp_export_bpf(filter, fileno(bpf_file));
    fflush(bpf_file);
    fclose(bpf_file);

    // Update the hash.
    if (arg::at("verbose") < "error" && arg::get("shell") != "debug") {
      auto hash = shared::hash(content[1] + arg::get("seccomp"));
      auto cache_file = std::ofstream(syscall_file);
      cache_file << "HASH: " << hash << "\n";
      cache_file << content[1];
      cache_file.close();
    }

    if (arg::at("stats")) {
      // x86 Syscalls
      float start = 335.0f, final = syscalls.size();
      std::cout << "Reduced Total Syscalls By: " << ((start - final) / fabs(start)) * 100.0f << "%"
                << " (" << start << " to " << final << ")" << std::endl;
    }

    // Release the filter and return the path to the filter so the main process can open it.
    seccomp_release(filter);
    return bpf;
  }


  // Update the policy based on strace output.
  void update_policy(const std::string& application, const vector &straced) {

    // Get our files.
    auto syscall_file = app_data / "syscalls.txt";
    set syscalls;
    if (std::filesystem::exists(syscall_file)) {
      auto content = file::parse<vector>(syscall_file, vectorize);
      switch (content.size()) {
        case 1:  if (!content[0].starts_with("HASH")) container::split<set, char>(syscalls, content[0], ' ', false); break;
        case 2: container::split<set, char>(syscalls, content[1], ' '); break;
        default: break;
      }
    }

    log({"Existing syscalls:", std::to_string(syscalls.size())});

    for (const auto& line : straced) {
      if (line.contains('(')) {
        auto start = line.find("] ");
        if (start == std::string::npos) continue;
        start += 2;

        auto end = line.find('(', start);
        if (end == std::string::npos) continue;

        auto syscall = line.substr(start, end - start);
        if (seccomp_syscall_resolve_name(syscall.c_str()) != __NR_SCMP_ERROR && !syscalls.contains(syscall)) {
          log({"New syscall:", syscall});
          syscalls.emplace(syscall);
        }
      }
    }

    auto joined = join(syscalls, ' ');
    auto out = std::ofstream(syscall_file);
    out << "HASH: " << 'P' << shared::hash(joined) << '\n' << joined;
    out.close();
  }
}
